prov/util/src/util_mem_monitor.c

```diff
@@ -251,18 +281,34 @@ void ofi_monitors_del_cache(struct ofi_mr_cache *cache)
 	}
 
 	pthread_mutex_unlock(&mm_lock);
+	pthread_rwlock_unlock(&mm_rwlock);
 }
 
-/* Must be called holding monitor lock */
+/* Must be called with locks in place like following
+ *	pthread_rwlock_rdlock(&mm_rwlock);
+ *	pthread_mutex_lock(&mm_lock);
+ *	ofi_monitor_notify();
+ *	pthread_mutex_unlock(&mm_lock);
+ *	pthread_rwlock_unlock(&mm_rwlock);
+ */
 void ofi_monitor_notify(struct ofi_mem_monitor *monitor,
-			const void *addr, size_t len)
+			const void *addr, size_t len, int defer)
 {
 	struct ofi_mr_cache *cache;
 
 	dlist_foreach_container(&monitor->list, struct ofi_mr_cache,
 				cache, notify_entries[monitor->iface]) {
 		ofi_mr_cache_notify(cache, addr, len);
 	}
+
+	if (!defer) {
+		dlist_foreach_container(&monitor->list, struct ofi_mr_cache,
+					cache, notify_entries[monitor->iface]) {
+			pthread_mutex_unlock(&mm_lock);
+			ofi_mr_cache_flush(cache, false);
+			pthread_mutex_lock(&mm_lock);
+		}
```

Comment:
I thought it'd be simpler this way in regards to managing locks but I can change it back if you prefer.

Reply:
In general, I dislike having a function release a lock that an external function has acquired.  It's messy and leads to races.  But I also don't see a clean way to avoid it with the mr cache.

I also looked at trying to limit the scope of the new rwlock to being used immediately around accessing the cache list.  But I couldn't do that cleanly either.

Reply:
Yeah I tried to get rwlock to be used around the cache list only but I realized that it would result in deadlocks with the mm_lock unless it encapsulates it.

Reply:
I'm thinking rwlock can be used around the dlist_foreach loop now that I've removed the mm_locking in the add_cache and delete_cache functions